---
title: "pubicData_plasma"
author: "Melanie Smith"
date: "04/03/2021"
output: word_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    message = FALSE,
    warning = FALSE,
    cache = FALSE,
    fig.align = "center",
    results = "hide"
)
```

```{r UDF}
`%notin%` <- Negate(`%in%`)
# function (x) to count the number of non-zero records in each column (ie per sample)
nonzero <- function(x) sum(x != 0)
```

# Load required libraries

```{r loadLibrary}

# tidyverse core packages
library(tibble)
library(plyr)
library(reshape2)
library(dplyr)
library(tidyr)
library(readr)
library(stringr)
library(ggplot2)
library(tidyverse)

# tidyverse-friendly packages
library(plotly)
library(ggrepel)
library(GGally)
library(tidyHeatmap)
library(magrittr)
library(ggpubr)
library(ggalt)

# other packages
library(stringr)
library(edgeR)
library(pander)
library(RColorBrewer)
library(viridis)
library(pheatmap)
library(RUVSeq)
library(preprocessCore)
library(dendextend)
library(sva)
library(quantro)
library(caret)
library(pastecs)
library(psych)
library(scales) # For percent_format()
library(here)


```

## Import GSE105052

```{r import GSE105052}

# import the raw counts
GSE105052_counts <- read.table(file = here("publicData/GSE105052_counts_rawData.txt"),
                        sep = "\t",
                        header = TRUE,
                        row.names = 1)

# make a simple metadata file
GSE105052_meta <- data.frame(samplename = colnames(GSE105052_counts)) %>% 
  dplyr::mutate(., copy = samplename) %>% 
  tidyr::separate(., col = copy, into = c("condition", "ID"), sep = "\\_") %>% 
  dplyr::mutate_if(., is.factor, as.character)

# quick scatterplot of counts
ggplot(data = as.data.frame(colSums(GSE105052_counts)) %>% 
         tibble::rownames_to_column() %>%
         set_colnames(c("samplename", "sampleSum")) %>%
         left_join(., GSE105052_meta, by = "samplename") %>%
         dplyr::select(., samplename, sampleSum)) +
  geom_point(aes(x = samplename, y = sampleSum, colour = GSE105052_meta$condition)) +
  geom_hline(yintercept = 1000000) +
  theme_bw() +
  ggtitle("Raw library size - GSE105052 plasma counts") +
  theme(axis.text.x = element_text(angle = 270, hjust = 1))


```

## Plot number of unique miRs as a function of readcounts

```{r unique miRs by read counts}

# start with a bit of exploration
# rank the samples by read counts and by unique miRs
rank <- as.data.frame(colSums(GSE105052_counts)) %>%
  set_colnames(., "readCounts") %>% 
                arrange(., -(readCounts)) %>% 
  tibble::rownames_to_column("samplename") %>% 
  left_join(., GSE105052_meta, by = "samplename") %>% 
  dplyr::select(., samplename, readCounts, condition) %>% 
  dplyr::mutate(., rank_readCounts = 1:nrow(.)) %>% 
  full_join(.,
            as.data.frame(t(numcolwise(nonzero)(as.data.frame(GSE105052_counts)))) %>%
              tibble::rownames_to_column() %>%
              set_colnames(., c("samplename", "unique_miRs")) %>%
              arrange(., desc(unique_miRs)) %>%
              mutate(., rank_unique = 1:nrow(.)),
            by = "samplename")

ggplot(data = rank,
       aes(x = readCounts,
           y = unique_miRs,
           size = 4)
       ) +
  scale_size(guide = "none") +
  geom_point() +
  scale_x_continuous(name = "Filtered Read Counts",
                     breaks = seq(1000000, 19000000, 1500000)) +
  scale_y_continuous(name = "Mature miRNA Identified",
                     breaks = seq(250, 800, 50)) +
  stat_smooth(method = 'loess',
              se = FALSE,
              size = 2) +
  geom_text_repel(data = dplyr::filter(rank, readCounts < 1000000),
                  aes(label = samplename)) +
  ggtitle("GSE105052: Mature miRNA as a function of Read Counts") +
  theme_bw(base_size = 16) +
  theme(axis.text.x = element_text(angle = 270, hjust = 1))

```


```{r remove small libraries}

# # identify samples with < 1 million reads
# lowCounts <- names(GSE105052_counts[, colSums(GSE105052_counts) < 1000000])
# 
# # remove columns/samples with readcouns less than 1 million
# GSE105052_counts <- GSE105052_counts[, colSums(GSE105052_counts) > 1000000]
# 
# # subset the metadata to only include samples that remain in the expression data
# combined_samples <- subset(combined_samples, samplename %in% colnames(GSE105052_counts))

```

# Filter sequencing noise   

* Here any individual observation with less than five counts is considered to be sequencing noise and is reduced to zero and removed by filtering.  
  * Counts less than 5 reduced to 0
  * Logic check and record how many miRs will be filtered
  * Filter to remove miRs with zero counts in all samples

```{r filterSeqNoise}

# reduce any individual count less than five to zero
GSE105052_counts[GSE105052_counts < 5] <- 0
# check to establish if any miRNA have no expression across all samples
table(rowSums(GSE105052_counts == 0) == ncol(GSE105052_counts))
  
# remove miRNAs with zero counts in all samples
GSE105052_counts <- GSE105052_counts[ rowSums(GSE105052_counts)!=0, ]

is.na(GSE105052_counts) %>% table()

```

## Establish DGEList object

```{r establish DGEList object}

# establish a DGEList object

DGEList_GSE105052 <- DGEList(counts = GSE105052_counts,
                             samples = GSE105052_meta)

```

## Quick MDS plot

```{r raw MDS}

# basic MDS plot
plotMDS(DGEList_GSE105052$counts)
title(main = "GSE105052 Plasma: Multi Dimensional Scaling Plot \n(raw data)")

```

# TMM normalisation of libraries

Before differential expression analysis the sample libraries need to be normalised to account for differences in initial library size.     
Normalising the libraries allows for the direct comparison between samples.   
Here the Trimmed Mean of M Values method is used.   

```{r TMM}

# Plot the distribution of (low count filtered) counts prior to normalisation 
nsamples <- ncol(DGEList_GSE105052)
# set up colours for the density plots
col <- colorRampPalette(brewer.pal(11,"Spectral"))(ncol(DGEList_GSE105052))
lcpm <- cpm(DGEList_GSE105052, log = TRUE)
boxplot(lcpm, las = 2, col = col, main = "")
title(main = "DGEList_GSE105052: Un-normalised data (n=42)",ylab = "Log-cpm") 

# calculate normalisation factors and apply to the DGEList object
DGEList_GSE105052 <- calcNormFactors(DGEList_GSE105052, method = "TMM")

```

# Filter biological noise and plot
 
```{r filter bionoise}

# calculate the cpm value of a read count of 40 given the mean library size (to be used in keep.genes)
cpm(40, mean(DGEList_GSE105052$samples$lib.size))

# new df of unfiltered cpm for the reduced DGEList
rawCpm_plasma <- cpm(DGEList_GSE105052, log = FALSE)

# new df of unfiltered log 2 cpm for the reduced DGEList
rawlcpm_plasma <- cpm(DGEList_GSE105052, log = TRUE)

## The density of log-CPM values for pre-filtered data 
# (A) and post-filtered data 
# (B) are shown for each sample. 
## Dotted vertical lines mark the log-CPM of 1 threshold 
# (equivalent to a CPM value of 2) used in the filtering step.
nsamples <- ncol(DGEList_GSE105052)

par(mfrow=c(1,1))

# colour by condition
col.condition <- DGEList_GSE105052$samples$condition
levels(col.condition) <- c("Patient", "Control")
col.condition <- as.character(col.condition)

plot(density(rawlcpm_plasma[,1]),
     col = ifelse(col.condition[1] == "Patient",'red','darkGrey'),
     lwd = 2,
     ylim = c(0, 0.80),
     las = 2,
  main = "",
  xlab = "")
title(main = "DGEList_GSE105052 Unfiltered data",
      xlab = "Log-cpm")
abline(v = 1, lty = 3)

for (i in 1:nsamples){
  den <- density(rawlcpm_plasma[,i])
  lines(den$x, den$y,
        col = ifelse(col.condition[i] == "Patient",'red','darkGrey'),
        lwd = 2)
}

# remove low expressed genes
keep.exprs <- rowSums(rawCpm_plasma > 40) >= 17
DGEList_GSE105052 <- DGEList_GSE105052[keep.exprs,, keep.lib.sizes = FALSE]

filtered_lcpm <- cpm(DGEList_GSE105052, log=TRUE)

plot(density(filtered_lcpm[,1]),
     col = ifelse(col.condition[1] == "Patient",'red','darkGrey'),
     lwd = 2,
     ylim = c(0, 0.20),
     las = 2,
  main = "", xlab = "")
title(main = "DGEList_GSE105052 Filtered data (20cpm)", xlab = "Log-cpm") 
abline(v = 1, lty = 3)
for (i in 1:nsamples){
  den <- density(filtered_lcpm[,i])
lines(den$x,
      den$y,
      col = ifelse(col.condition[i] == "Patient",'red','darkGrey'),
      lwd = 2)
}

par(mfrow=c(1,1))
# Distribution of normalised and filtered counts data
boxplot(cpm(DGEList_GSE105052, log = TRUE), las = 2, col = col, main = "")
title(main = "DGEList_GSE105052: Normalised data",ylab = "Log-cpm")

```

## MDS of filtered and TMM normalised counts

```{r mds filtered and normalised}

# calculate MDS data
MDS <- plotMDS(DGEList_GSE105052, main = "MDS norm counts")

# pull out x and y
x <- MDS$x
y <- MDS$y  

# plot with ggplot
cbind(x, y) %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("samplename") %>% 
  full_join(., DGEList_GSE105052$samples, by = "samplename") %>% 
  ggplot() +
  geom_point(aes(x = x,
                 y = y,
                 colour = condition,
                 size = 6)) +
  geom_text(aes(x = x,
                y = y,
                label = samplename,
                colour = condition), vjust = -1) +
  labs(title = "MDS plot normalised counts - GSE105052 \nplasma",
       y = "DIM 2",
       x = "DIM 1") +
  theme_bw()

```

## calculate the difference between the geometric mean of the distributions for
  - here we calculate the geometric mean of the classifier distribution and the
  "other" ensuring those taken from the classifier list are not included in
  other.

```{r calculate median difference reduced classifier list}

# create a vector of sample names for use in the lapply
varc <- dplyr::select(DGEList_GSE105052$samples, samplename) %>%
  tibble::remove_rownames() %>% 
  dplyr::pull(., samplename)

classifier_miRs <- data.frame(
  SYMBOL = c(
    "hsa-miR-106b-3p",
    "hsa-miR-140-3p",
    "hsa-miR-142-5p",
    "hsa-miR-532-5p",
    "hsa-miR-17-5p",
    "hsa-miR-19b-3p",
    "hsa-miR-30c-5p",
    "hsa-miR-324-5p",
    "hsa-miR-192-5p",
    "hsa-miR-660-5p",
    "hsa-miR-186-5p",
    "hsa-miR-425-5p",
    "hsa-miR-25-3p",
    "hsa-miR-363-3p",
    "hsa-miR-183-5p",
    "hsa-miR-451a",
    "hsa-miR-182-5p",
    "hsa-miR-191-5p",
    "hsa-miR-194-5p",
    "hsa-miR-20b-5p"
  ))

project_miRs <- data.frame(SYMBOL = c("NA"))

# define the dropped classifiers
dropped <- subset(classifier_miRs, SYMBOL %in% project_miRs$SYMBOL)

# define the final set of classifiers
final_classifiers <- subset(classifier_miRs, SYMBOL %notin% project_miRs$SYMBOL)

# calculate the distribution difference between the final classifier miRs (ie after removing pregnancy associated)
distributionDifference <- lapply(varc, function(x){
  # calculate the geometric mean of the two distributions (1 = classifier, 0 = other, 2 = dropped) 
  cdat <- ddply(
    dplyr::select(
      as.data.frame(cpm(DGEList_GSE105052$counts, log = TRUE)), x) %>%
      tibble::rownames_to_column("mirna") %>% 
      mutate(., classifier = as.factor(ifelse(mirna %in% final_classifiers$SYMBOL, 1,
                                              ifelse(mirna %in% dropped$SYMBOL, 2,
                                                     ifelse(mirna %notin% classifier_miRs$SYMBOL, 0, NA))))),
    "classifier", summarise, geometric.mean = psych::geometric.mean(get(x), na.rm = TRUE)
  )
  
  # calculate the difference between the two geometric means (classifier-other)  
  dplyr::filter(cdat, classifier == 1)$geometric.mean - dplyr::filter(cdat, classifier == 0)$geometric.mean
  
})

names(distributionDifference) <- varc

unlist_distributionDifference <- do.call(cbind.data.frame, distributionDifference) %>% 
  t() %>%
  set_colnames("distributionDifference") %>%
  as.data.frame() %>% 
  tibble::rownames_to_column("samplename") %>% 
  dplyr::mutate(., haemoResult = ifelse(distributionDifference < 1.9, "Clear",
                                        ifelse(distributionDifference >= 1.9, "Caution", NA)))

# import the distribution difference data from my samples
plotData_distDiff_dCq <- read.table(here("shinyVamp/inputData/plotData_distDiff_dCq.csv"),
                                    header = TRUE,
                                    sep = ",")

# plot as density side by side
ggplot() +
  geom_density(data = plotData_distDiff_dCq,
    aes(
      x = distributionDifference,
      fill = haemolysis,
      colour = haemolysis
      ),
    alpha = 0.5,
    lwd = 0.8
    ) +
   geom_density(data = unlist_distributionDifference,
    aes(
      x = distributionDifference),
    alpha = 0.5,
    lwd = 0.8,
    fill = "grey"
    ) +
  labs(
    title = "Distribution difference using final classifiers",
    subtitle = "based on three classification groups"
    ) +
  theme_bw(base_size = 16)

# plot as histogram side by side
p <- ggplot() +
  geom_histogram(data = plotData_distDiff_dCq,
    aes(x = distributionDifference,
    fill = haemolysis,
    colour = haemolysis,
      y = 2*(..density..)/sum(..density..)),
    breaks = seq(0,5,0.1),
    alpha = 0.6, 
    position = "identity",
    lwd = 0.8) +
  geom_histogram(
    data = unlist_distributionDifference,
    aes(x = distributionDifference,
        fill = "grey",
        colour = "grey",
        y = 2*(..density..)/sum(..density..)),
    breaks = seq(0,5,0.1),
    alpha = 0.6, 
    position = "identity",
    lwd = 0.8) +
  geom_vline(show.legend = FALSE,
             xintercept = 1.9,
             col = 2,
             lty = 2) +
  scale_y_continuous(labels = percent_format()) +
  labs(
    title = "Distribution difference using final classifiers",
    subtitle = "based on three classification groups"
    ) +
  theme_bw(base_size = 16)

caution <- length(filter(unlist_distributionDifference, haemoResult == "Caution"))

p + annotate(
    geom = "text",
    x = 3,
    y = .25,
    label = paste("we have identified", caution, "samples to use with caution", sep = " "),
    colour = "red"
  )

ggplot(data = left_join(rank, unlist_distributionDifference, by = "samplename"),
       aes(x = readCounts,
           y = unique_miRs)) +
  geom_point(aes(colour = haemoResult,
                 size = 4)) +
    scale_size(guide = "none") +
  scale_x_continuous(name = "Filtered Read Counts",
                     breaks = seq(min(rank$readCounts), max(rank$readCounts), 100000)) +
  scale_y_continuous(name = "Mature miRNA Identified",
                     breaks = seq(250, 800, 50)) +
  stat_smooth(method = 'loess',
              se = FALSE,
              size = 2) +
  geom_text_repel(data = filter(left_join(rank, unlist_distributionDifference, by = "samplename"), haemoResult == "Caution"),
    box.padding = 1,
    aes(label = samplename),
    show.legend = FALSE) +
  ggtitle("GSE105052: Mature miRNA as a function of Read Counts") +
  theme_bw(base_size = 16) +
  theme(axis.text.x = element_text(angle = 270, hjust = 1))

```

